src/client/action/list_databases.rs

```diff
@@ -0,0 +1,168 @@
+use std::{future::IntoFuture, marker::PhantomData};
+
+use bson::{Bson, Document};
+use futures_util::FutureExt;
+
+#[cfg(any(feature = "sync", feature = "tokio-sync"))]
+use crate::sync::Client as SyncClient;
+use crate::{
+    client::BoxFuture,
+    error::{ErrorKind, Result},
+    operation::list_databases as op,
+    results::DatabaseSpecification,
+    Client,
+    ClientSession,
+};
+
+impl Client {
+    /// Gets information about each database present in the cluster the Client is connected to.
+    pub fn list_databases(&self) -> ListDatabases {
+        ListDatabases {
+            client: &self,
+            options: Default::default(),
+            session: None,
+            mode: PhantomData,
+        }
+    }
+
+    /// Gets the names of the databases present in the cluster the Client is connected to.
+    pub fn list_database_names(&self) -> ListDatabases<'_, Names> {
+        ListDatabases {
+            client: &self,
+            options: Default::default(),
+            session: None,
+            mode: PhantomData,
+        }
+    }
+}
+
+#[cfg(any(feature = "sync", feature = "tokio-sync"))]
+impl SyncClient {
+    /// Gets information about each database present in the cluster the Client is connected to.
+    pub fn list_databases(&self) -> ListDatabases {
+        self.async_client.list_databases()
+    }
+}
+
+/// Gets information about each database present in the cluster the Client is connected to.
+#[must_use]
+pub struct ListDatabases<'a, M: Mode = Full> {
```

Comment:
Have you thought at all about combining this type with the `operation::ListDatabases` type (and ditto for future action types we introduce)? It would require some refactoring to store the client/session as part of the operation's state, but it might end up making the operation layer a little bit more streamlined if the compiler allows it...

Reply:
That's a really interesting possibility.  I'll definitely give this a try and see if I can make the refactor work.

Reply:
Unfortunately, there are some significant hurdles to this approach :(  I've filed RUST-1816 for this - it _might_ be possible but it'd be a huge refactor of the executor, change streams, and cursors.